eggtimer.js

```diff
@@ -1,326 +1,681 @@
+        })
+        .catch((err) => {
+            console.log("No PR is in the merging state: ", err);
+            currentContext = null;
+        })
+        .finally(() => {
+            console.log("Start listening on " + Config.port);
+            http.createServer((req, res) => {
+                 WebhookHandler(req, res, () => {
+                   res.statusCode = 404;
+                   res.end('no such location');
+                 });
+            }).listen(Config.port);
+            if (currentContext === null)
+                processNextPR(true);
+        });
+}
+
+function signaled() {
+    return (currentContext !== null && currentContext.signaled);
+}
+
+function processEvent() {
+    if (currentContext !== null) {
+        console.log("Signaling PR" + currentContext.pr.number, currentContext.pr.head.sha);
+        currentContext.signaled = true;
+    } else {
+        processNextPR(true);
+    }
+}
+
+function processNextPR(all) {
+    const sig = signaled();
+    currentContext = null;
+    if (all) {
+        getPRList();
+        return;
+    } else if (PRList.length === 0 && !sig) {
+        console.log("No more PRs to process.");
+        return;
+    } else if (sig) {
+        getPRList();
+        return;
     }
+    let prContext = createPRContext(PRList.shift());
+    console.log("Processing PR" + prContext.pr.number, prContext.pr.head.sha);
+    currentContext = prContext;
+
+    checkMergePreconditions(prContext);
+}
+
+function mergingTag(prNum) {
+    return "tags/" + MergingTag + prNum;
+}
+
+function checkMergePreconditions(prContext) {
+    let reviewsParams = commonParams();
+    reviewsParams.number = prContext.pr.number;
+    let reviewsPromise = getReviews(reviewsParams);
+
+    let statusParams = commonParams();
+    statusParams.ref = prContext.pr.head.sha;
+    let statusPromise = getStatuses(statusParams);
+
+    let mergeableParams = commonParams();
+    mergeableParams.number = prContext.pr.number;
+    let mergeablePromise = getPR(mergeableParams);
+
+    let labelsParams = commonParams();
+    labelsParams.number = prContext.pr.number;
+    let labelsPromise = getLabels(labelsParams);
+
+    Promise.all([reviewsPromise, statusPromise, mergeablePromise, labelsPromise]).then((results) => {
+            if (prContext.signaled)
+                return Promise.reject("signaled");
+
+            let reviews = results[0];
+            let statuses = results[1];
+            let mergeable = results[2].mergeable;
+            let labels = results[3];
+
+            if ((mergeable !== true) || !approved(reviews) || !allChecksSuccessful(statuses))
+                return Promise.reject("not ready for merging yet");
+
+            if (markedAsMerged(labels))
+                return Promise.reject("already merged");
+
+            if (markedAsFailed(labels))
+                console.log("PR" + prContext.pr.number + " previous merge was unsuccessful");
+
+            checkBeforeStartMerging(prContext);
```

Comment:
This result-ignoring checkBeforeStartMerging() call that is not followed by some kind of merge() call is very confusing! I eventually found startMerging() inside checkBeforeStartMerging() but that is not an acceptable design. We should probably be doing some .then.then chaining here. These problems may be related to the reject-on-success weirdness inside checkBeforeStartMerging().

Reply:
I believe I have already documented what confused me.

Yes, the function is misnamed, but I doubt that adding AndStart suffix to the name is the best or even correct solution here because I suspect that this necessity to add AndFooAndBarAndBaz suffixes is an overall design problem, not a naming problem. With the right design, this naming problem would not exist.